lib/hashie/extensions/strict_key_access.rb

```diff
@@ -0,0 +1,71 @@
+module Hashie
+  module Extensions
+    # SRP: This extension will fail an error whenever a key is accessed that does not exist in the hash.
+    #
+    #   EXAMPLE:
+    #
+    #     class StrictKeyAccessHash < Hash
+    #       include Hashie::Extensions::StrictKeyAccess
+    #     end
+    #
+    #     >> hash = StrictKeyAccessHash[foo: "bar"]
+    #     => {:foo=>"bar"}
+    #     >> hash[:foo]
+    #     => "bar"
+    #     >> hash[:cow]
+    #       KeyError: key not found: :cow
+    #
+    # NOTE: For googlers coming from Python to Ruby, this extension makes a Hash behave like a "Dictionary".
+    #
+    module StrictKeyAccess
+      class DefaultError < StandardError
+        def initialize(msg = 'Setting or using a default with Hashie::Extensions::StrictKeyAccess does not make sense', *args)
+          super
+        end
+      end
+
+      # NOTE: This extension would break the default behavior of Hash initialization:
+      #
+      #     >> a = StrictKeyAccessHash.new(a: :b)
+      #     => {}
+      #     >> a[:a]
+      #       KeyError: key not found: :a
+      #
+      # Includes the Hashie::Extensions::MergeInitializer extension to get around that problem.
+      # Also note that defaults still don't make any sense with a StrictKeyAccess.
+      def self.included(base)
+        # Can only include into classes with a hash initializer
+        base.send(:include, Hashie::Extensions::MergeInitializer)
```

Comment:
No, this still seems completely unrelated to me. It looks like you're expecting a basic `Hash.new(a: :b)` to initialize the created hash to `{:a => :b}`, which is not the expected behavior.

When you pass any object to `Hash.new` (including a preexisting Hash), the result is a Hash that has a default value of the object you passed in. In my experience, this trips up a lot of people, but it doesn't mean that it's not the default behavior.

```ruby
test = Hash.new(a: :b)  #=> {}
```

In the case of a Hash with `StrictKeyAccess` mixed in, passing a default value into `Hash.new` doesn't semantically make sense, but neither does the result of implicitly mixing in `MergeInitializer`. When I mix in `StrictKeyAccess`, I want strict key access ... not strict key access AND merge initialization. Doing things implicitly, in my experience, leads to confusion for users because it adds behavior they weren't expecting.

In an ideal world, I would want to disallow passing in the default value when `StrictKeyAccess` is mixed in, but I'm not sure how well that would play with the other mixins. I would need to play around with it a bit. As the extension currently stands -- sans the implicit inclusion of `MergeInitializer` -- this implementation works as expected:

```ruby
class Dictionary < Hash
  include Hashie::Extensions::StrictKeyAccess
end

# Note: This is just for example. I don't think you should _ever_ do this next line
# without something like MergeInitializer because the default implementation is
# not what you'd think.
test = Dictionary.new(a: :b)  # => {}
test[:a]  #=> KeyError: key not found: :a
```

If the user _wants_ merge initialization, that's fine -- they can also mix in `MergeInitializer`. That's why we have it after all. :smiley:

I would rather this extension _not_ also include `MergeInitializer`, since `MergeInitializer` has nothing to do with `StrictKeyAccess`.